services/libs/integrations/src/integrations/twitter/processStream.ts

```diff
@@ -0,0 +1,189 @@
+  const data = ctx.stream.data as TwitterMentionsStreamData
+
+  const { records, nextPage, limit, timeUntilReset } = await getPostsByMention(
+    {
+      token: ctx.integration.token,
+      page: data.page,
+      perPage: 100,
+      profileId: ctx.integration.identifier,
+    },
+    ctx,
+  )
+
+  const sleep = limit <= 1 ? timeUntilReset : undefined
+
+  if (sleep) {
+    throw new RateLimitError(sleep, 'twitter/mentions')
+  }
+
+  if (records && records.length && records.length > 0) {
+    for (const rec of records) {
+      await ctx.publishData<TwitterPublishData>({
+        type: TwitterStreamType.MENTIONS,
+        data: rec,
+      })
+    }
+  }
+
+  if (nextPage) {
+    await ctx.publishStream<TwitterMentionsStreamData>(
+      `${TwitterStreamType.MENTIONS}:${nextPage}`,
+      {
+        page: nextPage,
+      },
+    )
+  }
+}
+
+const processHashtagStream: ProcessStreamHandler = async (ctx) => {
+  const data = ctx.stream.data as TwitterHashtagStreamData
+
+  const { records, nextPage, limit, timeUntilReset } = await getPostsByHashtag(
+    {
+      token: ctx.integration.token,
+      page: data.page,
+      perPage: 100,
+      hashtag: data.hashtag,
+    },
+    ctx,
+  )
+
+  const sleep = limit <= 1 ? timeUntilReset : undefined
+
+  if (sleep) {
+    throw new RateLimitError(sleep, 'twitter/hashtag')
+  }
+
+  if (records && records.length && records.length > 0) {
+    for (const rec of records) {
+      await ctx.publishData<TwitterPublishData>({
+        type: TwitterStreamType.HASHTAG,
+        data: rec,
+        hashtag: data.hashtag,
+      })
+    }
+  }
+
+  if (nextPage) {
+    await ctx.publishStream<TwitterHashtagStreamData>(`${TwitterStreamType.HASHTAG}:${nextPage}`, {
+      page: nextPage,
+      hashtag: data.hashtag,
+    })
+  }
+}
+
+const processReachStream: ProcessStreamHandler = async (ctx) => {
+  const data = ctx.stream.data as TwitterReachStreamData
+  const usernames = data.usernames
+
+  if (!usernames) {
+    // this is the initial stream, we need to get all the usernames
+    const perPage = 100
+    const db = ctx.getDbConnection()
+
+    ctx.log.info('Getting all usernames for reach update', { int: ctx.integration })
+
+    await processPaginated<ReachSelection>(
+      async (page) => {
+        const result = await db.any(
+          `
+          SELECT
+            m."memberId" as id,
+            m.username as username
+          FROM
+            "memberIdentities" m
+          WHERE
+            m."tenantId"= (select "tenantId" from integrations where id = $(integrationId) )
+            and m.platform = $(platform)
+          ORDER BY
+            m."memberId"
+          LIMIT $(perPage)
+          OFFSET $(offset)
+        `,
+          {
+            integrationId: ctx.integration.id,
+            platform: PlatformType.TWITTER,
+            perPage,
+            offset: (page - 1) * perPage,
+          },
+        )
```

Comment:
nit: let's keep the queries in repo layer - what do u think?

Reply:
Yeah, but there is no repo layer in integrations folder - I can put this query just in a separate file inside integration if it makes sense

Reply:
I think if we at least keep the query in a separate function with a name to what it does it'll be cleaner. Not a strong opinion tho